File: backend/app/schemas/students.py
"""Student schemas for IMKON LMS — matching imkon-payment structure."""

from datetime import date

from pydantic import BaseModel, ConfigDict, computed_field

from .base import PaginatedList


class StudentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    document_id: str
    first_name: str
    last_name: str
    middle_name: str | None = None
    full_name: str = ""  # from model property via from_attributes
    birth_date: date | None = None
    gender: str | None = None
    phone_number: str | None = None
    photo_url: str | None = None
    role: str
    is_active: bool
    student_id: str | None = None
    grade_id: int | None = None
    father_first_name: str | None = None
    father_last_name: str | None = None
    father_phone: str | None = None
    mother_first_name: str | None = None
    mother_last_name: str | None = None
    mother_phone: str | None = None
    address: str | None = None
    enrollment_date: date | None = None
    withdrawal_date: date | None = None

    # Freeze fields
    is_frozen: bool = False
    frozen_at: date | None = None
    frozen_reason: str | None = None
    departure_date: date | None = None
    return_date: date | None = None

    # Soft delete fields
    is_deleted: bool = False
    deleted_at: date | None = None

    @computed_field
    @property
    def father_full_name(self) -> str | None:
        if not self.father_last_name and not self.father_first_name:
            return None
        parts = [p for p in [self.father_last_name, self.father_first_name] if p]
        return " ".join(parts) if parts else None

    @computed_field
    @property
    def mother_full_name(self) -> str | None:
        if not self.mother_last_name and not self.mother_first_name:
            return None
        parts = [p for p in [self.mother_last_name, self.mother_first_name] if p]
        return " ".join(parts) if parts else None


StudentList = PaginatedList[StudentRead]


class StudentFreezeRequest(BaseModel):
    departure_date: date | None = None
    reason: str | None = None


class StudentUnfreezeRequest(BaseModel):
    return_date: date


class StudentFreezeResponse(BaseModel):
    id: int
    full_name: str
    is_frozen: bool
    frozen_at: date | None = None
    frozen_reason: str | None = None
    departure_date: date | None = None
    return_date: date | None = None
    message: str

File: frontend/src/lib/api/students.ts
import { api } from "./client"

export interface StudentRead {
  id: number
  document_id: string
  first_name: string
  last_name: string
  middle_name: string | null
  full_name: string
  student_id: string | null
  grade_id: number | null
  birth_date: string | null
  gender: string | null
  phone_number: string | null
  photo_url: string | null
  father_first_name: string | null
  father_last_name: string | null
  father_phone: string | null
  father_full_name: string | null
  mother_first_name: string | null
  mother_last_name: string | null
  mother_phone: string | null
  mother_full_name: string | null
  address: string | null
  enrollment_date: string | null
  withdrawal_date: string | null
  is_active: boolean
  is_frozen: boolean
  frozen_at: string | null
  frozen_reason: string | null
  departure_date: string | null
  return_date: string | null
  is_deleted: boolean
  deleted_at: string | null
}

export interface StudentList {
  data: StudentRead[]
  count: number
}

export const studentsApi = {
  list: (
    params: {
      skip?: number
      limit?: number
      grade_id?: number
      search?: string
      status?: string
    } = {},
  ) => api.get<StudentList>("/api/v1/students/", { params }),
}
